File: src/test/java/oscar/cp/test/TestVariable.java
/*******************************************************************************
 * This file is part of OscaR (Scala in OR).
 *   
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *  
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/gpl-3.0.html
 ******************************************************************************/
package oscar.cp.test;


import junit.framework.TestCase;
import java.util.Arrays;
import java.util.Random;

import oscar.cp.constraints.*;
import oscar.cp.core.*;
import oscar.cp.search.*;
import oscar.cp.util.*;
import oscar.reversible.*;
import oscar.search.*;


/**
 * @author Pierre Schaus pschaus@gmail.com
 */
public class TestVariable extends TestCase {

	private Store s;
	private CPVarInt a;
	private CPVarInt b;
	private CPVarInt c;


	public TestVariable(String name) {
		super(name);

	}

	/**
	 * setUp() method that initializes common objects
	 */
	protected void setUp() throws Exception {
		super.setUp();
		s = new Store();
		a = CPVarInt.apply(s,1,6);
		b = CPVarInt.apply(s,0,31);
		c = CPVarInt.apply(s,1,10);
	}

	/**
	 * tearDown() method that cleanup the common objects
	 */
	protected void tearDown() throws Exception {
		super.tearDown();
		s = null;
		a = null;
		b = null;
		c = null;
	}




	public void testa(){


		assertTrue(true);

		a.removeValue(2);
		a.removeValue(4);

		s.pushState();

		a.removeValue(2);
		a.removeValue(10);
		a.removeValue(6);

		s.pushState();

		a.removeValue(-2);
		a.removeValue(4);
		a.removeValue(1);
		a.removeValue(4);

		s.pop();

		a.removeValue(1);
		a.removeValue(0);

		s.pop();

		//[1, 3, 5, 6] card:4 min:1 max:6
		assertTrue(a.getMin()==1);
		assertTrue(a.getMax()==6);
		assertTrue(a.getSize()==4);
		assertTrue(a.hasValue(1));
		assertTrue(a.hasValue(3));
		assertTrue(a.hasValue(5));
		assertTrue(a.hasValue(6));
		assertTrue(!a.hasValue(2));

	}

	public void testb(){

		b.assign(0);
		assertTrue(b.isBound() && b.getValue()==0);
	}

	public void testc(){

		c.removeValue(5);
		c.updateMax(7);
		//1..4,6..7
		int i = 0;
		for (int v = c.min(); v < c.max(); v++) {
			if (c.hasValue(v)) {
				switch (i) {
				case 0:
					assertTrue(v == 1);
					break;
				case 1:
					assertTrue(v == 2);
					break;
				case 2:
					assertTrue(v == 3);
					break;
				case 3:
					assertTrue(v == 4);
					break;
				case 4:
					assertTrue(v == 6);
					break;
				case 5:
					assertTrue(v == 7);
					break;
				default:
					break;
				}
				i++;
			}
		}
	}


	public void testd(){
		Store cp = new Store();
		CPVarInt x = CPVarInt.apply(cp,1,6);
		assertTrue(x.valueAfter(5) == 6);
		assertTrue(x.valueAfter(-10) == 1);
		
		int v  = x.valueAfter(6);
		assertTrue(v == 6);


		CPVarInt y = CPVarInt.apply(cp,-100,100);
		y.removeValue(0);


		assertTrue(y.valueAfter(-1) == 1);
		assertTrue(y.valueAfter(0) == 1);
		assertTrue(y.valueBefore(-1) == -2);
		assertTrue(y.valueBefore(1) == -1);
		assertTrue(y.valueBefore(0) == -1);
		assertTrue(y.valueBefore(1000) == 100);

		y.removeValue(30);
		y.removeValue(31);
		y.removeValue(32);
		y.removeValue(33);

		assertTrue(y.valueBefore(31) == 29);
		assertTrue(y.valueBefore(34) == 29);
		assertTrue(y.valueAfter(31) == 34);
		assertTrue(y.valueAfter(34) == 35);     

	}


	public void teste(){
		Store cp = new Store();
		int [] freq = new int[4]; 
		CPVarInt x = CPVarInt.apply(cp,new int[]{0,1,2,3});
		for (int i = 0; i < 200; i++) {
			freq[x.randomValue()]++;
		}
		for (int i = 0; i < 4; i++) {
			assertTrue(freq[i]>0);
		}
		System.out.println(Arrays.toString(freq));
	}

	public void testf(){
		Store cp = new Store();
		CPVarInt x = CPVarInt.apply(cp,new int[]{1,5,9,10});
		CPVarInt y = CPVarInt.apply(cp,new int[]{5,9,11});
		CPVarInt z = CPVarInt.apply(cp,new int[]{6,7,11});
		CPVarInt w = CPVarInt.apply(cp, 14);
		
		assertTrue(x.intersectionSize(y) == 2);
		
		assertTrue(y.intersectionSize(x) == 2);

		assertTrue(z.intersectionSize(y) == 1);
		assertTrue(y.intersectionSize(z) == 1);

		assertTrue(w.intersectionSize(x) == 0);
		assertTrue(x.intersectionSize(w) == 0);
	}


}

File: oscar-cp/src/test/scala/oscar/cp/test/TestVariable.scala
package oscar.cp.test

import org.scalatest.FunSuite
import org.scalatest.matchers.ShouldMatchers
import oscar.cp.core.CPStore
import oscar.cp.core.CPVarInt

class TestVariable extends FunSuite with ShouldMatchers {

  trait StoreAndVariables {
    val store = new CPStore()
    val a = CPVarInt(1, 6)(store)
    val b = CPVarInt(0, 31)(store)
    val c = CPVarInt(1, 10)(store)
  }

  test("test a") {
    new StoreAndVariables {
      a.removeValue(2)
      a.removeValue(4)
      store.pushState()
      a.removeValue(2)
      a.removeValue(10)
      a.removeValue(6)
      store.pushState()
      a.removeValue(-2)
      a.removeValue(4)
      a.removeValue(1)
      a.removeValue(4)
      store.pop()
      a.removeValue(1)
      a.removeValue(0)
      store.pop()
      //[1, 3, 5, 6] card:4 min:1 max:6
      assert(a.min == 1)
      assert(a.max == 6)
      assert(a.size == 4)
      assert(a.hasValue(1))
      assert(a.hasValue(3))
      assert(a.hasValue(5))
      assert(a.hasValue(6))
      assert(!a.hasValue(2))
    }
  }

  test("test b") {
    new StoreAndVariables {
      b.assign(0)
      assert(b.isBound && b.value == 0)
    }
  }

  test("test c") {
    new StoreAndVariables {
      c.removeValue(5)
      c.updateMax(7)
      //1..4,6..7
      var values = List(1, 2, 3, 4, 6, 7)
      for (v <- c.min until c.max) {
        if (c.hasValue(v)) {
          assert(v == values.head)
          values = values.tail
        }
      }
    }
  }

  test("test d") {
    val cp = new CPStore()
    val x = CPVarInt(1, 6)(cp)

    assert(x.valueAfter(5) == 6)
    assert(x.valueAfter(-10) == 1)
    val v: Int = x.valueAfter(6)
    assert(v == 6)

    val y = CPVarInt(-100, 100)(cp)
    y.removeValue(0)

    assert(y.valueAfter(-1) == 1)
    assert(y.valueAfter(0) == 1)
    assert(y.valueBefore(-1) == -2)
    assert(y.valueBefore(1) == -1)
    assert(y.valueBefore(0) == -1)
    assert(y.valueBefore(1000) == 100)

    y.removeValue(30)
    y.removeValue(31)
    y.removeValue(32)
    y.removeValue(33)

    assert(y.valueBefore(31) == 29)
    assert(y.valueBefore(34) == 29)
    assert(y.valueAfter(31) == 34)
    assert(y.valueAfter(34) == 35)
  }

  // Non determinist test !
  test("test e") {
    val cp = new CPStore()
    val freq = Array.fill(4)(0)
    val x = CPVarInt(Set(0, 1, 2, 3))(cp)
    for (i <- 0 until 200) {
      freq(x.randomValue) += 1
    }
    for (i <- 0 until 4) {
      assert(freq(i) > 0)
    }
    println(freq.mkString(", "))
  }

  test("test f") {
    val cp = new CPStore()
    val x = CPVarInt(Set(1, 5, 9, 10))(cp)
    val y = CPVarInt(Set(5, 9, 11))(cp)
    val z = CPVarInt(Set(6, 7, 11))(cp)
    val w = CPVarInt(14)(cp)

    assert(x.intersectionSize(y) == 2)
    assert(y.intersectionSize(x) == 2)

    assert(z.intersectionSize(y) == 1)
    assert(y.intersectionSize(z) == 1)

    assert(w.intersectionSize(x) == 0)
    assert(x.intersectionSize(w) == 0)
  }
}
